compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -60,35 +60,43 @@ object JavaNullInterop {
     if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
       // Don't nullify the `TYPE` field in every class and Java enum instances
       tp
-    else if (sym.name == nme.toString_ || sym.isConstructor)
-      // Don't nullify the return type of the `toString` method and constructors
-      nullifyParamsOnly(tp)
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNullAnnot(sym))
+      // Don't nullify the return type of the `toString` method.
+      // Don't nullify the return type of constructors.
+      // Don't nullify the return type of methods with a not-null annotation.
+      nullifyExceptReturnType(tp)
     else
       // Otherwise, nullify everything
       nullifyType(tp)
   }
 
-  /** Only nullify method parameters (but not result types). */
-  private def nullifyParamsOnly(tp: Type)(implicit ctx: Context): Type =
-    new JavaNullMap(alreadyNullable = false, nonNullResultType = true)(ctx)(tp)
+  private def hasNotNullAnnot(sym: Symbol)(implicit ctx: Context): Boolean =
+    ctx.definitions.NotNullAnnots.exists(nna => sym.unforcedAnnotation(nna).isDefined)
+
+  /** If tp is a MethodType, the parameters and the inside of return type are nullified,
+   *  but the result return type is not nullable.
+   *  If tp is a type of a field, the inside of the type is nullified,
+   *  but the result type is not nullable.
+   */
+  private def nullifyExceptReturnType(tp: Type)(implicit ctx: Context): Type =
+    new JavaNullMap(true)(ctx)(tp)
 
   /** Nullifies a Java type by adding `| JavaNull` in the relevant places. */
   private def nullifyType(tp: Type)(implicit ctx: Context): Type =
-    new JavaNullMap(alreadyNullable = false, nonNullResultType = false)(ctx)(tp)
+    new JavaNullMap(false)(ctx)(tp)
 
 
   /** A type map that adds `| JavaNull`.
-   *  @param alreadyNullable whether the type being mapped is already nullable (at the outermost level).
-   *                         This is needed so that `JavaNullMap(A | B)` gives back `(A | B) | JavaNull`,
-   *                         instead of `(A|JavaNull | B|JavaNull) | JavaNull`.
-   *  @param nonNullResultType if true, then the current type is a method or generic method type,
-   *                           and we don't want to nullify its return type at the top level.
-   *                           This is useful e.g. for constructors.
+   *  @param currentLevelNullable whether the type needs to be nullable at current level.
```

Comment:
Hmm, this comments seems wrong, because if `currentLevelNullable = true`, then the type will _not_ be nullified at the outermost level.

I propose we rename this parameter to `outermostLevelAlreadyNullable`, and change the comment to

```Scala
/*
A type map that implements the nullification function on types. Given a Java-sourced type, this adds `| JavaNull`
in the right places to make the nulls explicit in Scala.

@param outermostLevelAlreadyNullable whether this type is already nullable at the outermost level.
                                                                 For example, `Array[String]|JavaNull` is already nullable at the outermost 
                                                                 level, but `Array[String|JavaNull]` isn't.
                                                                 If this parameter is set to true, then the types of fields, and the return types
                                                                 of methods will not be nullified.
                                                                 This is useful for e.g. constructors, and also so that `A & B` is nullified to 
                                                                 `(A & B) | JavaNull`, instead of `(A|JavaNull & B|JavaNull) | JavaNull`.
*/